tensorboard/plugins/core/core_plugin.py

```diff
@@ -166,25 +169,86 @@ def get_first_event_timestamp(run_name):
 
   @wrappers.Request.application
   def _serve_experiments(self, request):
-    """Serve a JSON array of experiments, ordered by experiment started time.
-    Sort order is by started time (aka first event time) with empty times sorted
-    last, and then ties are broken by sorting on the experiment name.
+    """Serve a JSON array of experiments. Experiments are ordered by experiment
+    started time (aka first event time) with empty times sorted last, and then
+    ties are broken by sorting on the experiment name.
     """
+    results = []
     if self._db_connection_provider:
       db = self._db_connection_provider()
       cursor = db.execute('''
         SELECT
+          experiment_id,
           experiment_name,
-          started_time IS NULL as started_time_nulls_last,
-          started_time
+          started_time,
+          started_time IS NULL as started_time_nulls_last
         FROM Experiments
-        ORDER BY started_time_nulls_last, started_time, experiment_name
+        ORDER BY started_time_nulls_last, started_time, experiment_name,
+            experiment_id
       ''')
-      experiments = [{'name': row[0]} for row in cursor]
-    else:
-      # experiment is only useful when using db.
-      experiments = []
-    return http_util.Respond(request, experiments, 'application/json')
+      results = [{
+        "id": row[0],
+        "name": row[1],
+        "startTime": row[2],
+      } for row in cursor]
+
+    return http_util.Respond(request, results, 'application/json')
+
+  @wrappers.Request.application
+  def _serve_experiment_runs(self, request):
+    """Serve a JSON runs of an experiment, specified with query param
+    `experiment`, with their nested data, tag, populated. Runs returned are
+    ordered by started time (aka first event time) with empty times sorted last,
+    and then ties are broken by sorting on the run name. Tags are sorted by
+    its inserted time, displayName, then name.
+    """
+    results = []
+    if self._db_connection_provider:
+      exp_id = request.args.get('experiment')
+      runs_dict = collections.defaultdict(dict)
```

Comment:
I don't think your code uses the fact that this is a defaultdict (since the logic that accesses it is all guarded by `if not run_id in runs_dict: runs_dict[run_id] = { ... }` so the default value creation never happens).

If you don't make it a defaultdict, you can make it an OrderedDict() instead and then omit `run_ids` and instead just do `results = list(runs_dict.values())`.